File: apps/server/agent/src/systemprompt/environment.ts
/**
 * Environment / workstation info for the system prompt.
 */
import { execFile } from "node:child_process";
import * as os from "node:os";
import * as path from "node:path";
import { promisify } from "node:util";
import type { EnvironmentInfo } from "@/agent/src/types/system-prompt.js";

const execFileAsync = promisify(execFile);

function formatLocalDate(date = new Date()): string {
  // YYYY-MM-DD in local timezone
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, "0");
  const d = String(date.getDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

async function getGitBranch(cwd: string): Promise<string | undefined> {
  try {
    const { stdout } = await execFileAsync("git", ["rev-parse", "--abbrev-ref", "HEAD"], {
      cwd,
      timeout: 2000,
      maxBuffer: 1024,
    });
    const branch = stdout.trim();
    return branch && branch !== "HEAD" ? branch : undefined;
  } catch {
    return undefined;
  }
}

export async function collectEnvironmentInfo(options: {
  cwd: string;
  model?: string;
}): Promise<EnvironmentInfo> {
  const cwd = path.resolve(options.cwd);
  const gitBranch = await getGitBranch(cwd);

  return {
    date: formatLocalDate(),
    cwd,
    os: `${os.platform()} ${os.release()}`,
    arch: os.arch(),
    node: process.version,
    gitBranch,
    model: options.model,
  };
}

File: apps/server-go/internal/agent/systemprompt/environment.go
// Environment / workstation info for the system prompt. Port of
// apps/server/agent/src/systemprompt/environment.ts.
package systemprompt

import (
	"context"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

// EnvironmentInfo is the host/environment summary shown in the workstation
// block of the system prompt.
type EnvironmentInfo struct {
	Date      string
	Cwd       string
	OS        string
	Arch      string
	GitBranch string
	Model     string
}

func gitBranch(cwd string) string {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	cmd := exec.CommandContext(ctx, "git", "rev-parse", "--abbrev-ref", "HEAD")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	branch := strings.TrimSpace(string(out))
	if branch == "" || branch == "HEAD" {
		return ""
	}
	return branch
}

// CollectEnvironmentInfo gathers date/cwd/OS/arch/git-branch for cwd.
func CollectEnvironmentInfo(cwd, model string) EnvironmentInfo {
	abs, err := filepath.Abs(cwd)
	if err != nil {
		abs = cwd
	}
	return EnvironmentInfo{
		Date:      time.Now().Format("2006-01-02"),
		Cwd:       abs,
		OS:        runtime.GOOS,
		Arch:      runtime.GOARCH,
		GitBranch: gitBranch(abs),
		Model:     model,
	}
}
